494 fe content panel bug tags

mitkapanarin opened the pull request:
Tag Management Requirements:

Both content managers and admins can add existing tags
Only admins can create new tags
All operations require edit mode

mitkapanarin committed:
fix: Edit Mode Restriction, Role-Based Tag Management

Edit Mode Restriction:  Tags can only be added when in edit mode Input and remove buttons are disabled outside edit mode Clear error message when attempting to add tags outside edit mode Role-Based Tag Management:  Both content managers and administrators can add existing tags Only administrators can create new tags that don't exist Different UI and messages based on user role

dichoskaemilija commented on apis/mutations/blogs/useAddNewPost.ts:
```diff
@@ -33,8 +33,12 @@ const useAddNewPost = () => {
     onSuccess: () => {
-      queryClient.invalidateQueries({ queryKey: QUERY_KEYS.BLOGS.ALL });
       toast.success('Статијата беше успешно објавена!');
+      onSuccess?.();
+      // Invalidate queries after showing success message
+      setTimeout(() => {
+        queryClient.invalidateQueries({ queryKey: QUERY_KEYS.BLOGS.ALL });
+      }, 100);
```
You dont need to call setTimeout inside onSucess. Follow the pattern bellow:
export const useAddNewPost = () => {
  const queryClient = useQueryClient();
  return useMutation((params: Params) => addNewPost(params), {
    onSuccess: (_, variables) => {
//Here the query key should be the one from the get blog posts query
      queryClient.invalidateQueries({ queryKey: ["newPost"] });
    },
  });

dichoskaemilija commented on apis/mutations/blogs/useEditBlogPost.ts:
```diff
@@ -45,8 +46,12 @@ const useEditBlogPost = () => {
+      // Invalidate queries after showing success message
+      setTimeout(() => {
+        queryClient.invalidateQueries({ queryKey: QUERY_KEYS.BLOGS.ALL });
+      }, 100);
```
same from above

dichoskaemilija commented on apis/mutations/blogs/useEditBlogPost.ts:
```diff
@@ -20,21 +20,22 @@ type ErrorResponse = {
-      const statusResponse = await axios.patch(`${ENDPOINTS.BLOGS.UPDATE_STATUS(id)}`, {
-        status,
-      });
+      const [editResponse, statusResponse] = await Promise.all([
+        axios.patch(ENDPOINTS.BLOGS.EDIT(id), {
+          title,
+          content,
+          tags,
+        }),
+        axios.patch(`${ENDPOINTS.BLOGS.UPDATE_STATUS(id)}`, {
+          status,
+        }),
+      ]);
```
what happens id the user doesn't update the status? we shouldnt make the second call in that scenario

dichoskaemilija commented on apis/mutations/blogs/useEditBlogPost.ts:
```diff
@@ -20,21 +20,22 @@ type ErrorResponse = {
 };
 
-const useEditBlogPost = () => {
+const useEditBlogPost = (onSuccess?: () => void) => {
```
Is there a reason we are accepting onSuccess as a parametar here?
the useEditBlogPost handles the onSuccess inside it. It returns isSuccess value.

dichoskaemilija commented on components/module-components/blog/PublishArticleForm.tsx:
```diff
@@ -1,21 +1,32 @@
+  const [selectedTags, setSelectedTags] = useState<Tag[]>([]);
+  const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
+
+  const addNewPostMutation = useAddNewPost(() => {
+    setHasUnsavedChanges(false);
+    router.push('/content-panel/blogs');
+  });
```
Here you can use the following pattern:
call the addNewPostMutation mutation, on Success handle the    
setHasUnsavedChanges(false);
router.push('/content-panel/blogs');

on Error display a toast with the error

dichoskaemilija commented on components/module-components/blog/TagInput.tsx:
```diff
@@ -12,20 +12,38 @@ import { Tag } from '../../reusable-components/_Types';
   const { data } = useGetTags(debouncedSearchTerm);
   const addNewTagMutation = useAddNewTag();
 
+  useImperativeHandle(ref, () => ({
+    clearInput: () => setSearchTag(''),
+  }));
+
```
What is the purpose of useImperativeHandle?

mitkapanarin replied:
Removed useImperativeHandle by lifting searchTag state to TagManager for better state management

dichoskaemilija reviewed (changes requested):
Great start! Just a few improvements suggested.
Can you please go through all the changed files and remove the comments 🙏

mitkapanarin committed:
refactor: refactor

refactor: simplify blog post mutation with direct query invalidation

mitkapanarin committed:
refactor: Refactor useEditBlogPost hook to remove unnecessary callback parameter

(useEditBlogPost): Remove onSuccess callback parameter and optimize API calls  (BlogDetailsCard): Update to use React Query's onSuccess option directly  (useEditBlogPost): Add conditional status update to avoid unnecessary API calls  (BlogDetailsCard): Add check for edit mode in handleBackClick condition

dichoskaemilija commented on components/module-components/blog/PublishArticleForm.tsx:
```diff
@@ -1,21 +1,32 @@
+  const [selectedTags, setSelectedTags] = useState<Tag[]>([]);
+  const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
+
+  const addNewPostMutation = useAddNewPost(() => {
+    setHasUnsavedChanges(false);
+    router.push('/content-panel/blogs');
```
here you can handle onSuccess and onError and return appropriate toast messages.
So you can follow the following pattern:
First define the mutation: const {mutate: addNewPost}= useAddNewPost()
So onSuccess => do this:
  const addNewPostMutation = (arguments) => {

    useAddNewPost(
      {
       arguments
      },
      {
        onSuccess: () => {
    setHasUnsavedChanges(false);
    router.push('/content-panel/blogs');
        },
        onError: (error: any) => {
    //show a toast 
        },
      },
    );
  };
  
 **This way the mutation hook will remain reusable and the side effects are localized**

dichoskaemilija commented on components/module-components/blog/PublishArticleForm.tsx:
```diff
@@ -31,119 +42,189 @@ const PublishArticleForm = () => {
+  }, [hasUnsavedChanges]);
+
   const handleAddPost = (values: NewPost) => {
+    setHasUnsavedChanges(false); // Reset immediately since we're about to save
```
remove comment

mitkapanarin committed:
feat: Add toasts and refactor mutation handling for new posts

mitkapanarin committed:
refactor: refactor(Tags): Replace useImperativeHandle with state lifting

Refactored TagManager and TagInput components. Removed the use of useImperativeHandle for clearing the search input.  Instead, the `searchTag` state is now managed by the parent TagManager and passed down to TagInput along with a change handler and a clear function. This aligns better with standard React data flow patterns.

dichoskaemilija reviewed (approved):
great work @mitkapanarin 👏
